File: desktop/macos/ShellProtocol.swift
import Foundation

/// Transient command pushed from Swift to the chat WebView.
enum ShellCommand: Equatable {
    case openCatalog
    case closeCatalog
    case openSkillsCatalog
    case closeSkillsCatalog
    case openCron(id: String)
    case openSettings
    case focusChat
}

/// Product action sent from the chat WebView to the native shell.
enum ShellAction: Equatable {
    case selectSession(id: String?)
    case createSession
    case archiveSession(id: String)
    case unarchiveSession(id: String)
    case renameSession(id: String, title: String)
    case sessionMutated(id: String)
    case sessionStreaming(id: String, streaming: Bool)
    case sessionUnread(id: String, unread: Bool)
    case cronsChanged
    case connectionsChanged
    case skillsChanged
    case openExternalUrl(url: String)
    case signOut
    case catalogClosed
    case skillsCatalogClosed
    case settingsVisibility(open: Bool)
}

/// Strict decoder for the untyped WKScriptMessage boundary. Required fields
/// fail closed so malformed web messages never become partial native actions.
enum ShellActionDecoder {
    static func decode(_ payload: [String: Any]) -> ShellAction? {
        guard let kind = payload["kind"] as? String else { return nil }
        switch kind {
        case "select-session":
            guard let rawID = payload["id"] else { return nil }
            if rawID is NSNull { return .selectSession(id: nil) }
            guard let id = rawID as? String else { return nil }
            return .selectSession(id: id)
        case "create-session":
            return .createSession
        case "archive-session":
            guard let id = payload["id"] as? String else { return nil }
            return .archiveSession(id: id)
        case "unarchive-session":
            guard let id = payload["id"] as? String else { return nil }
            return .unarchiveSession(id: id)
        case "rename-session":
            guard let id = payload["id"] as? String,
                  let title = payload["title"] as? String else { return nil }
            return .renameSession(id: id, title: title)
        case "session-mutated":
            guard let id = payload["id"] as? String else { return nil }
            return .sessionMutated(id: id)
        case "session-streaming":
            guard let id = payload["id"] as? String,
                  let streaming = payload["streaming"] as? Bool else { return nil }
            return .sessionStreaming(id: id, streaming: streaming)
        case "session-unread":
            guard let id = payload["id"] as? String,
                  let unread = payload["unread"] as? Bool else { return nil }
            return .sessionUnread(id: id, unread: unread)
        case "crons-changed":
            return .cronsChanged
        case "connections-changed":
            return .connectionsChanged
        case "skills-changed":
            return .skillsChanged
        case "open-external-url":
            guard let url = payload["url"] as? String else { return nil }
            return .openExternalUrl(url: url)
        case "sign-out":
            return .signOut
        case "catalog-closed":
            return .catalogClosed
        case "skills-catalog-closed":
            return .skillsCatalogClosed
        case "settings-visibility":
            guard let open = payload["open"] as? Bool else { return nil }
            return .settingsVisibility(open: open)
        default:
            return nil
        }
    }
}

File: desktop/chat-ui/src/shell-protocol.ts
// Wire format between the Swift shell and the chat-ui WebView.
//
// All product-level shell IPC uses three channels:
//
//   • Swift → JS: `verso:shell-state` carrying a full `ShellState` snapshot.
//   • Swift → JS: `verso:shell-command` for transient commands.
//   • JS → Swift: `chatBridge.postMessage({type: 'action', action})` with a
//     single discriminated `ShellAction` payload.
//
// `desktop/macos/ChatWebView.swift` contains the matching Swift definitions.

import type { ChatSessionSummary } from './types';

/// Snapshot of everything the chat-ui's UI derives from. Pushed from Swift
/// after every mutation that affects what the chat-ui should render.
/// Last write wins.
export interface ShellState {
  sessions: ChatSessionSummary[];
  selectedSessionId: string | null;
}

/// Transient command pushed from Swift to JS. Not snapshot-able (overlay
/// open/close state lives in the chat-ui).
export type ShellCommand =
  | { kind: 'open-catalog' }
  | { kind: 'close-catalog' }
  | { kind: 'open-skills-catalog' }
  | { kind: 'close-skills-catalog' }
  | { kind: 'open-cron'; id: string }
  | { kind: 'open-settings' }
  | { kind: 'focus-chat' };

/// Action sent from JS → Swift via the chatBridge.
export type ShellAction =
  | { kind: 'select-session'; id: string | null }
  | { kind: 'create-session' }
  | { kind: 'archive-session'; id: string }
  | { kind: 'unarchive-session'; id: string }
  | { kind: 'rename-session'; id: string; title: string }
  // "I just streamed a message into session X; please refresh." Used so
  // Swift's leftbar picks up the AI-generated title that lands after the
  // first response.
  | { kind: 'session-mutated'; id: string }
  // Streaming state changed for a session. Swift's leftbar uses this to
  // show a "working" indicator on rows whose agent is currently generating,
  // so the user can tell which conversations are alive when they're not
  // looking at the chat surface.
  | { kind: 'session-streaming'; id: string; streaming: boolean }
  // Unread response for a session — set when a stream finishes while the
  // user wasn't looking at that chat surface (different session selected,
  // or an overlay covered chat). Cleared when the chat-ui sees the session
  // become actively-viewed again. Chat-ui owns the rule because only it
  // knows full overlay state.
  | { kind: 'session-unread'; id: string; unread: boolean }
  | { kind: 'crons-changed' }
  | { kind: 'connections-changed' }
  | { kind: 'skills-changed' }
  | { kind: 'open-external-url'; url: string }
  | { kind: 'sign-out' }
  // User dismissed the catalog via the chat-ui's close button (rather
  // than via a Swift-side leftbar toggle).
  | { kind: 'catalog-closed' }
  | { kind: 'skills-catalog-closed' }
  // Settings page visibility. The native shell hides the sessions sidebar
  // while settings is open so the settings rail is the window's only left
  // column; the chat-ui owns navigation state, so it reports transitions.
  | { kind: 'settings-visibility'; open: boolean };
